### ISSUE TITLE
Entity with a Projection as Property

### ISSUE THREAD

**Issue by KaNarlist** [dotnet/efcore#23308]
Is there an easy way to have an entity with a projection as child property?

I have a Table A and a Table B. A has two foreign keys to B. So I have EntityA and EntityB that both represent their table with EntityA having two references of EntityB

```
    class EntityA
    {
        public int FirstBId { get; set; }
        public EntityB FirstB { get; set; }
        public int SecondBId { get; set; }
        public EntityB SecondB { get; set; }
    }
```

But now instead of EntityB I want to use another class that is a projection of B with only a couple of Bs members.

```
    class EntityA
    {
        public int FirstBId { get; set; }
        public ProjectionOfB FirstB { get; set; }
        public int SecondBId { get; set; }
        public ProjectionOfB SecondB { get; set; }
    }
```

The only way to accomblish this I could get to work this far is setting FirstB and SecondB to Ignore in the configuration of EntityA, joining EntityA with EntityB two times and then setting both to the instance of EntityA manually after everything is loaded. That's obviously pretty... unpretty. Is there an easier way to do this that I am missing?

```
var result = Context.EntityA.Join(Context.EntityB.Select(b => new ProjectionOfB() { ...}),
                                                    m => m.FirstBId, 
						    e => e.Id,
						    (m, e) => new { EntityA = m, FirstB = e })
                                            .Join(Context.EntityB.Select(b => new ProjectionOfB() { ...}),
						    m => m.EntityA.SecondBId,
                                                    e => e.Id,
                                                   (m, e) => new { m.EntityA, m.FirstB, SecondB = e }).ToList();
								
foreach(var r in result)
{
	r.EntityA.FirstB = r.FirstB;
	r.EntityA.SecondB = r.SecondB;
}
```

**Comment by ajcvickers** [comment]
@KaNarlist In the code above there are two definitions of `EntityA`. Are they in different namespaces? Is one included in the EF model but not the other?

**Comment by KaNarlist** [comment]
No it's just an example. The first one how it is working and the second how I want it to work.

**Comment by ajcvickers** [comment]
@KaNarlist I still don't fully understand. You can use an anonymous type for the projection. For example:

```C#
var results1 = context.EntityAs.Select(e => new
{
    e.FirstBId,
    FirstB = new ProjectionOfB
    {
        Id = e.FirstB.Id,
    },
    e.SecondBId,
    SecondB = new ProjectionOfB
    {
        Id = e.FirstB.Id,
    },
}).ToList();
```

Or you can create explicit types for the projection:

```C#
public class ProjectionOfB
{
    public int Id { get; set; }
}

public class ProjectionOfA
{
    public int FirstBId { get; set; }
    public ProjectionOfB FirstB { get; set; }
    public int SecondBId { get; set; }
    public ProjectionOfB SecondB { get; set; }
}
```

And use those:

```C#
var results2 = context.EntityAs.Select(e => new ProjectionOfA
{
    FirstBId = e.FirstBId,
    FirstB = new ProjectionOfB
    {
        Id = e.FirstB.Id,
    },
    SecondBId = e.SecondBId,
    SecondB = new ProjectionOfB
    {
        Id = e.FirstB.Id,
    },
}).ToList();
```

But I don't understand what it would mean to have an entity type directly reference a projection.

**Comment by KaNarlist** [comment]
I have a complex database that is configured through the fluent api with complex entities referencing other complex entities. So EntityA from my example would reference several other entities where some also have child entities and so on. For one of those complex entites in EntityA I now want a simple class now instead that only contains the most relevant data from the originaly referenced entity and its childs. 

The only way to do this I see so far is either by doing a Select(x => new EntityA() {...}) and constructing the complete EntityA in the Select statement manually despite allready having it configured with the fluent api and the other way is joining it into a temporary result and then manually putting it together.

I guess what I am searching for is something like configuring a Projection through the fluent api somehow, so I can connect to it in the configuration of EntityA and everything gets loaded automatically.

**Comment by ajcvickers** [comment]
@KaNarlist You might want to look into [AutoMapper](https://www.nuget.org/packages/AutoMapper/). It can be used to configure projections.

**Comment by KaNarlist** [comment]
Thanks for the tipp. I already thought about it, but don't want to add the overhead of having another mapping.

I had another idea to solve it by creating a database view for the projection and map this to an entity. I didn't think of this before because usally I have to take the database as it is, but adding a view here wouldn't be a problem.
So I made a view and mapped ProjectionOfB to it with the fluent api. It seems to work fine when loading EntityA, so that all fields of ProjectionOfB are also loaded properly.
But updating EntityA fails when EF Core tries to update the view because obviously it doesn't know its just a view. I saw that I can make Keyless entities that are only queried but I can't used that because then I can't setup the realtions in EntityA to ProjectionOfB.

EF tries to update ProjectionOfB because this is all done in a multi-tier client/server application where EntityA is disconnected from DbContext when it is sent from a client. So it is added to DbContext with the Update() Method which sets the whole graph to modified or added. This is done because I want the entire graph to be updated in the db accordingly, just not the projections. So I am searching for some way now to configure my projection entities as readonly through the fluent api.

**Comment by KaNarlist** [comment]
I found a way that seems to work. When I set an AfterSaveBehaviour to ignore changes on every property of ProjectionOfB, besides of the key, I can update EntityA and everything in ProjectionOfB is ignored. If I set the behaviour on the key too, ef crashes with an InvalidOperationException. But I think it's ok for now to leave the key without the behaviour and just don't change it.

It might be worth considering providing an easier solution for this. But for now I think I am good.
